galter_subjects_utils/updater.py

```diff
@@ -152,13 +198,58 @@ def at_least_1_subject_targeted(ops_data):
     return result
 
 
+@dataclass
+class KeepTrace:
+    """Keeps trace of subject at field in record using template."""
+
+    field: str  # dotted path to field
+    template: str
+
+    def trace(self, record, subject):
+        """Save expanded `self.template` at `self.field` in record."""
+        if not self.field or not self.template or not subject:
+            return
+
+        final_dict = self.find_final_dict(record)
+        self.assign_template(final_dict, subject)
+
+    def find_final_dict(self, record):
+        """Find or create final dict by following `field`."""
+        obj = record
+        keys = self.field.split(".")
+
+        for key in keys[:-1]:
+            got = obj.get(key)
+            if isinstance(got, dict):
+                obj = got
+            elif got is None:
+                new_dict = {}
+                obj[key] = new_dict
+                obj = new_dict
+            elif isinstance(got, list):
+                new_dict = {}
+                got.append(new_dict)
+                obj = new_dict
+            else:
+                break
+
+        assert isinstance(obj, dict), f"KeepTrace.field '{self.field}' is invalid."  # noqa
+        return obj
+
+    def assign_template(self, dict_, subject):
+        """Assign expanded template."""
+        final_key = self.field.split(".")[-1]
+        dict_[final_key] = self.template.format(subject=subject)
+
+
 class SubjectDeltaUpdater:
     """Translates delta operations into actual changes."""
 
-    def __init__(self, ops_data, logger):
+    def __init__(self, ops_data, logger, keep_trace):
```

Comment:
I know it's a small change, but seeing the change in full I think this was the right choice